core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct Pointer<T> {
+	ptr: u32,
+	_marker: PhantomData<T>,
+}
+
+impl<T> Pointer<T> {
+	/// Create a new instance of `Self`.
+	pub fn new(ptr: u32) -> Self {
+		Self {
+			ptr,
+			_marker: Default::default(),
+		}
+	}
+
+	/// Calculate the offset from this pointer.
+	///
+	/// `offset` is in units of `T`. So, `3` means `3 * mem::size_of::<T>()` as offset to the pointer.
+	pub fn offset(self, offset: u32) -> Self {
+		Self {
+			ptr: self.ptr + offset * mem::size_of::<T>() as u32,
+			_marker: Default::default(),
+		}
+	}
+
+	/// Create a null pointer.
+	pub fn null() -> Self {
+		Self::new(0)
+	}
+
+	/// Cast this pointer of type `T` to a pointer of type `R`.
+	pub fn cast<R>(self) -> Pointer<R> {
+		Pointer::new(self.ptr)
+	}
+}
+
+impl<T> From<Pointer<T>> for u32 {
+	fn from(ptr: Pointer<T>) -> Self {
+		ptr.ptr
+	}
+}
+
+impl<T> From<Pointer<T>> for usize {
+	fn from(ptr: Pointer<T>) -> Self {
+		ptr.ptr as _
+	}
+}
+
+impl<T> IntoValue for Pointer<T> {
+	const VALUE_TYPE: ValueType = ValueType::I32;
+	fn into_value(self) -> Value { Value::I32(self.ptr as _) }
+}
+
+impl<T> TryFromValue for Pointer<T> {
+	fn try_from_value(val: Value) -> Option<Self> {
+		match val {
+			Value::I32(val) => Some(Self::new(val as _)),
+			_ => None,
+		}
+	}
+}
+
+/// The word size used in wasm. Normally known as `usize` in Rust.
+pub type WordSize = u32;
+
+/// The Signature of a function
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Signature {
+	/// The arguments of a function.
+	pub args: Cow<'static, [ValueType]>,
+	/// The optional return value of a function.
+	pub return_value: Option<ValueType>,
+}
+
+impl Signature {
+	/// Create a new instance of `Signature`.
+	pub fn new<T: Into<Cow<'static, [ValueType]>>>(args: T, return_value: Option<ValueType>) -> Self {
+		Self {
+			args: args.into(),
+			return_value,
+		}
+	}
+
+	/// Create a new instance of `Signature` with the given `args` and without any return value.
+	pub fn new_with_args<T: Into<Cow<'static, [ValueType]>>>(args: T) -> Self {
+		Self {
+			args: args.into(),
+			return_value: None,
+		}
+	}
+
+}
+
+/// Something that provides a function implementation on the host for a wasm function.
+pub trait Function {
+	/// Returns the name of this function.
+	fn name(&self) -> &str;
+	/// Returns the signature of this function.
+	fn signature(&self) -> Signature;
+	/// Execute this function with the given arguments.
+	fn execute(
+		&self,
+		context: &mut dyn FunctionContext,
+		args: &mut dyn Iterator<Item=Value>,
+	) -> Result<Option<Value>, String>;
+}
+
+/// Context used by `Function` to interact with the allocator and the memory of the wasm instance.
+pub trait FunctionContext {
+	/// Read memory from `address` into a vector.
+	fn read_memory(&self, address: Pointer<u8>, size: WordSize) -> Result<Vec<u8>, String> {
+		let mut vec = Vec::with_capacity(size as usize);
+		unsafe { vec.set_len(size as usize); }
```

Comment:
This lacks the safety proof and it would be problematic to guarantee *here* that this is safe since this depends on the implementation of `read_memory_into`, namely:

1. that it doesn't read the provided buffer,
2. that it writes `size` bytes.

It's UB if the implementer of `FunctionContext` doesn't ensure these invariants. That means in turn that `read_memory_into` should be marked `unsafe`.

Are you sure that `unsafe` block is required here? I think chances are that the zero initialization will be actually elided.